src/main/scala/mesosphere/marathon/util/LifeCycledCloseable.scala

```diff
@@ -0,0 +1,39 @@
+package mesosphere.marathon
+package util
+
+import java.io.Closeable
+import scala.util.Try
+
+trait LifeCycledCloseableLike[T <: Closeable] extends Closeable {
+  def closeable: T
+  def beforeClose(fn: () => Unit): Unit
+  def removeBeforeClose(fn: () => Unit): Unit
+}
+
+class LifeCycledCloseable[T <: Closeable](val closeable: T) extends LifeCycledCloseableLike[T] {
+  @volatile private[this] var closed: Boolean = false
+  @volatile private[this] var beforeCloseHooks: List[() => Unit] = Nil
+
+  override def close(): Unit = {
+    val hooks = synchronized {
```

Comment:
Shouldn't we check if _closeable_ wasn't already closed?

Reply:
Maybe? What problem will it avoid? Closing is an idempotent operation. Calling `close` twice on a java.io.InputStream does not throw an exception.